components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt

```diff
@@ -63,10 +89,60 @@ object WebExtensionSupport {
             }
 
             override fun onInstalled(webExtension: WebExtension) {
+                installedExtensions.add(webExtension)
                 store.dispatch(WebExtensionAction.InstallWebExtension(webExtension.toState()))
+
+                // Register a global default action handler on the new extension
+                webExtension.registerActionHandler(object : ActionHandler {
+                    override fun onBrowserAction(
+                        webExtension: WebExtension,
+                        session: EngineSession?,
+                        action: BrowserAction
+                    ) {
+                        store.dispatch(WebExtensionAction.UpdateBrowserAction(webExtension.id, action))
+                    }
+                })
+
+                // Register action handler for all existing engine sessions on the new extension
+                store.state.tabs
+                    .forEach {
+                        val session = it.engineState.engineSession
+                        if (session != null) {
+                            registerSessionActionHandler(webExtension, session, SessionActionHandler(store, it.id))
+                        }
+                    }
             }
         })
     }
 
+    /**
+     * Observes the provided store to register session-specific [ActionHandler]s
+     * for all installed extensions on newly added sessions.
+     */
+    private fun registerActionHandlersForNewSessions(store: BrowserStore) {
```

Comment:
Could you help me to clarify `newly added sessions` term mean? Does that mean session that are just linked to its respective `engineSession`, right?

Reply:
Sure, we're already capturing the existing sessions when the extension is installed. So, for every session added after (once it's engine session is created / linked) we have to do this.

Reply:
Thanks 👍